File: core/src/main/kotlin/eldoria/core/model/SubRealm.kt
package eldoria.core.model

/** Underground dungeon (tunnels/caves) or the sky realm reached by beanstalk. */
enum class RealmKind { DUNGEON, SKY_REALM }

enum class QuestType { RETRIEVE_ARTIFACT, DEFEAT_GUARDIAN, RESCUE_CAPTIVE }

/** One room/chamber (dungeon) or island/terrace (sky realm) inside a SubRealm. */
data class SubRealmRoom(
    val id: String,
    val name: String,
    val description: String,
    val difficultyTier: Int,
    val isBossRoom: Boolean,
    val beings: List<SpawnEntry>,
    val items: List<Item>,
    /** direction ("north","south","east","west","up","down", or a fallback "passage_*") -> room id */
    val exits: Map<String, String>,
)

data class SubRealmQuest(
    val title: String,
    val type: QuestType,
    val objective: String,
    val questItem: Item,
    val legendaryItem: Item,
)

data class SubRealm(
    val id: String,
    val kind: RealmKind,
    val name: String,
    val biome: Biome,
    val entranceLocationId: String,
    val entryRoomId: String,
    val bossRoomId: String,
    val rooms: Map<String, SubRealmRoom>,
    val quest: SubRealmQuest,
)

File: eldoria/models/sub_realm.py
"""A dungeon or sky realm: a graph of rooms reachable through an overworld portal tile."""
from __future__ import annotations

from dataclasses import dataclass, field

from eldoria.models.biome import Biome
from eldoria.models.game_location import SpawnEntry
from eldoria.models.item import Item
from eldoria.models.realm import QuestType, RealmKind

__all__ = ["RealmKind", "QuestType", "SubRealmRoom", "SubRealmQuest", "SubRealm"]


@dataclass(frozen=True)
class SubRealmRoom:
    """One room/chamber (dungeon) or island/terrace (sky realm) inside a SubRealm."""

    id: str
    name: str
    description: str
    difficulty_tier: int
    is_boss_room: bool
    beings: tuple[SpawnEntry, ...]
    items: tuple[Item, ...]
    exits: dict[str, str]


@dataclass(frozen=True)
class SubRealmQuest:
    title: str
    type: QuestType
    objective: str
    quest_item: Item
    legendary_item: Item


@dataclass(frozen=True)
class SubRealm:
    id: str
    kind: RealmKind
    name: str
    biome: Biome
    entrance_location_id: str
    entry_room_id: str
    boss_room_id: str
    rooms: dict[str, SubRealmRoom]
    quest: SubRealmQuest
